### PR TITLE
[4.0] Get the MVCFactory trough the ComponentInterface

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for Issue #17705.

### Summary of Changes
A common use case is that extensions need models from a component. There is no standard way to do that. With this pr, the `ComponentInterface` gets a new function to get it's `MVCFactory`. Internally this is provided trough a new factory in the service provider.

### Testing Instructions
- Create a latest articles module on the front end
- Open the front end

### Expected result
Latest module shows the latest articles.

### Actual result
Latest module shows the latest articles.

### CODE REVIEW

File: libraries/src/Dispatcher/Dispatcher.php [hunk 1]
```diff
@@ -62,23 +62,34 @@
...
 		if (empty($this->namespace))
 		{
 			throw new \RuntimeException('Namespace can not be empty!');
 		}
 
-		$this->app   = $app;
-		$this->input = $input ?: $app->input;
+		$this->app        = $app;
+		$this->input      = $input ?: $app->input;
+		$this->mvcFactory = $mvcFactory;
```
Comment (wilsonge) [inline libraries/src/Dispatcher/Dispatcher.php]: If the plan is to make this null - there needs to be a default. Else this needs to be first parameter and compulsory

File: libraries/src/Dispatcher/Dispatcher.php [hunk 2]
```diff
@@ -205,8 +216,22 @@ public function getController(string $name, string $client = '', array $config =
 			throw new \InvalidArgumentException(\JText::sprintf('JLIB_APPLICATION_ERROR_INVALID_CONTROLLER_CLASS', $controllerClass));
 		}
 
-		$controller = new $controllerClass($config, new MVCFactory($namespace, $this->app), $this->app, $this->input);
+		$factory = null;
```
Comment (wilsonge) [inline libraries/src/Dispatcher/Dispatcher.php]: This isn't required with your if/else. Also just to be clear - which one are you planning on keeping and which to be removed?

File: libraries/src/Dispatcher/Dispatcher.php (hunk 1 above)
Comment (laoneo) [reply libraries/src/Dispatcher/Dispatcher.php]: Made it required in the last commit, together with the input.

File: libraries/src/Dispatcher/Dispatcher.php (hunk 2 above)
Comment (laoneo) [reply libraries/src/Dispatcher/Dispatcher.php]: Removed it as the injected factory is required now.

File: tests/unit/suites/libraries/cms/model/JModelFormTest.php [hunk 3]
```diff
@@ -34,14 +34,11 @@ public function setUp()
 	{
 		$this->saveFactoryState();
 
+		JFactory::$application = $this->getMockCmsApp();
+
 		// Create mock of abstract class JModelForm to test concrete methods in there
 		$this->object = $this->getMockBuilder('JModelForm')
 			->getMockForAbstractClass();
-		$mockApp = $this->getMockCmsApp();
```
Comment (wilsonge) [inline tests/unit/suites/libraries/cms/model/JModelFormTest.php]: What's the reason for this change?

Comment (laoneo) [reply tests/unit/suites/libraries/cms/model/JModelFormTest.php]: I was too lazy to mock the factory :speak_no_evil: . Changed that.

File: libraries/src/Dispatcher/Dispatcher.php [hunk 4]
```diff
@@ -62,23 +62,34 @@
...
 	protected $input;
 
+	/**
+	 * The MVC factory
+	 *
+	 * @var  MVCFactoryFactoryInterface
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private $mvcFactory;
```
Comment (wilsonge) [inline libraries/src/Dispatcher/Dispatcher.php]: `$mvcFactoryFactory` please - I got really confused on us calling `createFactory` on this thing further down until i carefully re-read the doc blocks here

File: libraries/src/MVC/Model/BaseDatabaseModel.php [hunk 5]
```diff
@@ -314,18 +313,10 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
-			$reflect = new \ReflectionClass($this);
-			if ($reflect->getNamespaceName())
-			{
-				// Guess the root namespace
-				$ns = explode('\\', $reflect->getNamespaceName());
-				$ns = implode('\\', array_slice($ns, 0, 3));
-
-				$factory = new MVCFactory($ns, \JFactory::getApplication());
-			}
+			$factory = Factory::getApplication()->bootComponent($this->option)->createMVCFactory(Factory::getApplication());
```
Comment (wilsonge) [inline libraries/src/MVC/Model/BaseDatabaseModel.php:316]: I don't think `$this->option` is namespace aware from a quick look at the code. This is probably going to give rubbish names (and also probably won't be prefixed with `com_`)

File: tests/unit/suites/libraries/cms/model/JModelItemTest.php [hunk 6]
```diff
@@ -33,8 +35,10 @@ class JModelItemTest extends TestCase
 	public function setUp()
 	{
 		// Create mock of abstract class JModelForm to test concrete methods in there
-		$this->object = $this->getMockBuilder('JModelItem')
-			->getMockForAbstractClass();
+		$this->object = $this->getMockForAbstractClass(
```
Comment (wilsonge) [inline tests/unit/suites/libraries/cms/model/JModelItemTest.php]: Can we use the builder still here please? PHPUnit is slowly deprecating all other approaches so I'd rather stay ahead of the curve here

File: tests/unit/suites/libraries/cms/model/JModelFormTest.php [hunk 7]
```diff
@@ -35,8 +37,10 @@ public function setUp()
 		$this->saveFactoryState();
 
 		// Create mock of abstract class JModelForm to test concrete methods in there
-		$this->object = $this->getMockBuilder('JModelForm')
-			->getMockForAbstractClass();
+		$this->object = $this->getMockForAbstractClass(
```
Comment (wilsonge) [inline tests/unit/suites/libraries/cms/model/JModelFormTest.php]: Same here - keep the builder just use the right methods on it please

File: tests/unit/suites/libraries/cms/model/JModelAdminTest.php [hunk 8]
```diff
@@ -33,8 +35,10 @@ class JModelAdminTest extends TestCase
 	public function setUp()
 	{
 		// Create mock of abstract class JModelAdmin to test concrete methods in there
-		$this->object = $this->getMockBuilder('JModelAdmin')
-			->getMockForAbstractClass();
+		$this->object = $this->getMockForAbstractClass(
```
Comment (wilsonge) [inline tests/unit/suites/libraries/cms/model/JModelAdminTest.php]: And here :)

File: modules/mod_articles_latest/mod_articles_latest.php [hunk 9]
```diff
@@ -12,6 +12,6 @@
 use Joomla\CMS\Helper\ModuleHelper;
 use Joomla\Module\ArticlesLatest\Site\Helper\ArticlesLatestHelper;
 
-$list = ArticlesLatestHelper::getList($params);
+$list = ArticlesLatestHelper::getList($params, $app->bootComponent('com_content')->createMVCFactory($app)->createModel('Articles', 'Site'));
```
Comment (wilsonge) [inline modules/mod_articles_latest/mod_articles_latest.php]: Aren't we going to end up with issues here because we're no longer able to insert the `array('ignore_request' => true)` param here - I think on list views that will mean this module breaks

File: libraries/src/MVC/Model/BaseDatabaseModel.php (hunk 5 above)
Comment (laoneo) [reply libraries/src/MVC/Model/BaseDatabaseModel.php:316]: It doesn't have to https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Extension/ExtensionManagerTrait.php#L43
